### PR TITLE
Call support

### PR DESCRIPTION
**Pull request by arugaz**

### CODE REVIEW

**Comment by ahmedRSA** [comment]
bro you said you can offer a call too
but i only see call reject code 

https://github.com/tulir/whatsmeow/issues/555#issuecomment-2015958442

**Comment by arugaz** [comment]
yes, need to test again havent commit yet

**Review by ahmedRSA — APPROVED** [review]
i tested call reject and this is working 
thank you for contribution

**Comment by gamersindo1223** [comment]
I've found a way to offer a call on dm
```// OfferCall offers a call to a user.
func (cli *Client) OfferCall(callTo types.JID, video bool) error {
	clientID := cli.getOwnJID()
	if clientID.IsEmpty() {
		return ErrNotLoggedIn
	}
	var offerLen uint8 = 6
	if video {
		offerLen++
	}
	callID := strings.ToUpper(hex.EncodeToString(random.Bytes(16)))
	plaintext, dsmPlaintext, err := marshalMessage(callTo, &waProto.Message{Call: &waProto.Call{CallKey: random.Bytes(32)}})
	if err != nil {
		return fmt.Errorf("failed to marshal call: %w", err)
	}
	destinationNode, includeIdentity := cli.encryptMessageForDevices(context.TODO(), []types.JID{clientID, callTo}, clientID, callID, plaintext, dsmPlaintext, nil)
	if includeIdentity {
		destinationNode = append(destinationNode, cli.makeDeviceIdentityNode())
	}
	offerContent := make([]waBinary.Node, 0, offerLen)
	offerContent = append(offerContent,
		waBinary.Node{Tag: "audio", Attrs: waBinary.Attrs{"enc": "opus", "rate": "16000"}},
		waBinary.Node{Tag: "audio", Attrs: waBinary.Attrs{"enc": "opus", "rate": "8000"}},
	)
	if video {
		offerContent = append(offerContent,
			waBinary.Node{Tag: "video", Attrs: waBinary.Attrs{"orientation": "0", "screen_width": "1080", "screen_height": "2340", "device_orientation": "0", "enc": "vp8", "dec": "vp8"}},
		)
	}
	offerContent = append(offerContent,
		waBinary.Node{Tag: "capability", Attrs: waBinary.Attrs{"ver": "1"}, Content: []byte{1, 4, 255, 131, 207, 4}},
		waBinary.Node{Tag: "destination", Content: destinationNode},
		waBinary.Node{Tag: "encopt", Attrs: waBinary.Attrs{"keygen": "2"}},
		waBinary.Node{Tag: "net", Attrs: waBinary.Attrs{"medium": "3"}},
	)
	return cli.sendNode(waBinary.Node{
		Tag:   "call",
		Attrs: waBinary.Attrs{"id": cli.GenerateMessageID(), "to": callTo},
		Content: []waBinary.Node{{
			Tag:     "offer",
			Attrs:   waBinary.Attrs{"call-id": callID, "call-creator": clientID},
			Content: offerContent,
		}},
	})
}```

**Comment by arugaz** [comment]
yeah, thats what was done 4 months ago, ill still looking for group call

**Comment by ahmedRSA** [comment]
> I've found a way to offer a call on dm
> 
> ```
> func (cli *Client) OfferCall(callTo types.JID, video bool) error {
> 	clientID := cli.getOwnJID()
> 	if clientID.IsEmpty() {
> 		return ErrNotLoggedIn
> 	}
> 	var offerLen uint8 = 6
> 	if video {
> 		offerLen++
> 	}
> 	callID := strings.ToUpper(hex.EncodeToString(random.Bytes(16)))
> 	plaintext, dsmPlaintext, err := marshalMessage(callTo, &waProto.Message{Call: &waProto.Call{CallKey: random.Bytes(32)}})
> 	if err != nil {
> 		return fmt.Errorf("failed to marshal call: %w", err)
> 	}
> 	destinationNode, includeIdentity := cli.encryptMessageForDevices(context.TODO(), []types.JID{clientID, callTo}, clientID, callID, plaintext, dsmPlaintext, nil)
> 	if includeIdentity {
> 		destinationNode = append(destinationNode, cli.makeDeviceIdentityNode())
> 	}
> 	offerContent := make([]waBinary.Node, 0, offerLen)
> 	offerContent = append(offerContent,
> 		waBinary.Node{Tag: "audio", Attrs: waBinary.Attrs{"enc": "opus", "rate": "16000"}},
> 		waBinary.Node{Tag: "audio", Attrs: waBinary.Attrs{"enc": "opus", "rate": "8000"}},
> 	)
> 	if video {
> 		offerContent = append(offerContent,
> 			waBinary.Node{Tag: "video", Attrs: waBinary.Attrs{"orientation": "0", "screen_width": "1080", "screen_height": "2340", "device_orientation": "0", "enc": "vp8", "dec": "vp8"}},
> 		)
> 	}
> 	offerContent = append(offerContent,
> 		waBinary.Node{Tag: "capability", Attrs: waBinary.Attrs{"ver": "1"}, Content: []byte{1, 4, 255, 131, 207, 4}},
> 		waBinary.Node{Tag: "destination", Content: destinationNode},
> 		waBinary.Node{Tag: "encopt", Attrs: waBinary.Attrs{"keygen": "2"}},
> 		waBinary.Node{Tag: "net", Attrs: waBinary.Attrs{"medium": "3"}},
> 	)
> 	return cli.sendNode(waBinary.Node{
> 		Tag:   "call",
> 		Attrs: waBinary.Attrs{"id": cli.GenerateMessageID(), "to": callTo},
> 		Content: []waBinary.Node{{
> 			Tag:     "offer",
> 			Attrs:   waBinary.Attrs{"call-id": callID, "call-creator": clientID},
> 			Content: offerContent,
> 		}},
> 	})
> }```
> ```

hi is this possible to send call for only like 5 second
or X second 1, 2second ?

**Comment by gamersindo1223** [comment]
no Idea

**Comment by brunoaduarte** [comment]
> I've found a way to offer a call on dm
> 
> ```
> func (cli *Client) OfferCall(callTo types.JID, video bool) error {
> 	clientID := cli.getOwnJID()
> 	if clientID.IsEmpty() {
> 		return ErrNotLoggedIn
> 	}
> 	var offerLen uint8 = 6
> 	if video {
> 		offerLen++
> 	}
> 	callID := strings.ToUpper(hex.EncodeToString(random.Bytes(16)))
> 	plaintext, dsmPlaintext, err := marshalMessage(callTo, &waProto.Message{Call: &waProto.Call{CallKey: random.Bytes(32)}})
> 	if err != nil {
> 		return fmt.Errorf("failed to marshal call: %w", err)
> 	}
> 	destinationNode, includeIdentity := cli.encryptMessageForDevices(context.TODO(), []types.JID{clientID, callTo}, clientID, callID, plaintext, dsmPlaintext, nil)
> 	if includeIdentity {
> 		destinationNode = append(destinationNode, cli.makeDeviceIdentityNode())
> 	}
> 	offerContent := make([]waBinary.Node, 0, offerLen)
> 	offerContent = append(offerContent,
> 		waBinary.Node{Tag: "audio", Attrs: waBinary.Attrs{"enc": "opus", "rate": "16000"}},
> 		waBinary.Node{Tag: "audio", Attrs: waBinary.Attrs{"enc": "opus", "rate": "8000"}},
> 	)
> 	if video {
> 		offerContent = append(offerContent,
> 			waBinary.Node{Tag: "video", Attrs: waBinary.Attrs{"orientation": "0", "screen_width": "1080", "screen_height": "2340", "device_orientation": "0", "enc": "vp8", "dec": "vp8"}},
> 		)
> 	}
> 	offerContent = append(offerContent,
> 		waBinary.Node{Tag: "capability", Attrs: waBinary.Attrs{"ver": "1"}, Content: []byte{1, 4, 255, 131, 207, 4}},
> 		waBinary.Node{Tag: "destination", Content: destinationNode},
> 		waBinary.Node{Tag: "encopt", Attrs: waBinary.Attrs{"keygen": "2"}},
> 		waBinary.Node{Tag: "net", Attrs: waBinary.Attrs{"medium": "3"}},
> 	)
> 	return cli.sendNode(waBinary.Node{
> 		Tag:   "call",
> 		Attrs: waBinary.Attrs{"id": cli.GenerateMessageID(), "to": callTo},
> 		Content: []waBinary.Node{{
> 			Tag:     "offer",
> 			Attrs:   waBinary.Attrs{"call-id": callID, "call-creator": clientID},
> 			Content: offerContent,
> 		}},
> 	})
> }```
> ```

That doesn't work

My client sends the package but nothing returns from the server

14:10:08.046 [Client/Send DEBUG]
<call id="3EB0096894E2F3735674B5" to="xxxxxxxxx@s.whatsapp.net">
  <offer call-creator="yyyyyyyyyy:18@s.whatsapp.net" call-id="585786509F7115EF059ED9AE799F23F8">
    <audio enc="opus" rate="16000"/>
    <audio enc="opus" rate="8000"/>
    <capability ver="1">0104ff83cf04</capability>
    <destination>
      <to jid="xxxxxxxxx@s.whatsapp.net"><enc type="pkmsg" v="2"></enc></to>
      <device-identity></device-identity>
    </destination>
    <encopt keygen="2"/>
    <net medium="3"/>
  </offer>
</call>

**Comment by ahmedRSA** [comment]
this stop working for some reason anyones having asame problem ?

**Comment by PurpShell** [comment]
change commit name to feature: call rejection function

**Comment by ahmedRSA** [comment]
will this merge ?

**Comment by mateusfmello** [comment]
> change commit name to feature: call rejection function

@PurpShell @tulir 

If I change the title of the PR, will you merge the feature?

I need this implementation and if @arugaz doesn't update the title, I'll implement it myself with a new title to merge into the project.

**Review by mateusfmello — APPROVED** [review]
I implemented the modifications and it works perfectly.

File: call.go [hunk 1]
```diff
@@ -87,3 +87,21 @@ func (cli *Client) handleCallEvent(node *waBinary.Node) {
 		cli.dispatchEvent(&events.UnknownCallEvent{Node: node})
 	}
 }
+
+// CallReject reject an incoming call.
+func (cli *Client) CallReject(callFrom types.JID, callID string) error {
```
Comment (tulir) [inline call.go:92]:
```suggestion
// RejectCall reject an incoming call.
func (cli *Client) RejectCall(callFrom types.JID, callID string) error {
```

**Pull request merged**